File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/profile/ReportUserSheet.kt
@file:OptIn(androidx.compose.material3.ExperimentalMaterial3Api::class)
@file:Suppress("LongMethod", "MagicNumber", "PackageNaming", "TooManyFunctions", "LongParameterList")

package app.pantopus.android.ui.screens.profile

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.SheetState
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.heading
import androidx.compose.ui.semantics.selected
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import app.pantopus.android.ui.components.GhostButton
import app.pantopus.android.ui.components.PrimaryButton
import app.pantopus.android.ui.theme.PantopusColors
import app.pantopus.android.ui.theme.PantopusTextStyle
import app.pantopus.android.ui.theme.Radii
import app.pantopus.android.ui.theme.Spacing

/**
 * Bottom-sheet host for [ReportUserSheetBody]. Wires the dismiss handler
 * and forwards a success-pulse so the parent screen can show its toast.
 */
@Composable
fun ReportUserSheet(
    userId: String,
    handle: String?,
    displayName: String,
    sheetState: SheetState,
    onDismiss: () -> Unit,
    onSubmitted: () -> Unit,
    viewModel: ReportUserSheetViewModel = hiltViewModel(),
) {
    val state by viewModel.state.collectAsStateWithLifecycle()
    val reason by viewModel.selectedReason.collectAsStateWithLifecycle()
    val details by viewModel.details.collectAsStateWithLifecycle()

    LaunchedEffect(state) {
        if (state is ReportSheetUiState.Succeeded) {
            onSubmitted()
        }
    }

    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = sheetState,
        containerColor = PantopusColors.appBg,
        modifier = Modifier.testTag("reportUserSheet"),
    ) {
        ReportUserSheetBody(
            handle = handle,
            displayName = displayName,
            state = state,
            selectedReason = reason,
            details = details,
            onSelectReason = viewModel::selectReason,
            onDetailsChange = viewModel::updateDetails,
            onSubmit = { viewModel.submit(userId) },
            onCancel = onDismiss,
        )
    }
}

/**
 * Render-only body for the Report-User sheet. Exposed for Paparazzi
 * baselines so the host's [ModalBottomSheet] isn't needed in tests.
 */
@Composable
fun ReportUserSheetBody(
    handle: String?,
    displayName: String,
    state: ReportSheetUiState,
    selectedReason: ReportReason?,
    details: String,
    onSelectReason: (ReportReason) -> Unit,
    onDetailsChange: (String) -> Unit,
    onSubmit: () -> Unit,
    onCancel: () -> Unit,
) {
    val isSubmitting = state is ReportSheetUiState.Submitting
    val detailsRequired = selectedReason == ReportReason.Other
    val canSubmit =
        selectedReason != null &&
            (!detailsRequired || details.trim().isNotEmpty()) &&
            !isSubmitting

    Column(
        modifier =
            Modifier
                .fillMaxWidth()
                .verticalScroll(rememberScrollState())
                .padding(horizontal = Spacing.s4)
                .padding(bottom = Spacing.s10),
        verticalArrangement = Arrangement.spacedBy(Spacing.s3),
    ) {
        Text(
            text = if (!handle.isNullOrEmpty()) "Report @$handle" else "Report $displayName",
            style = PantopusTextStyle.h3,
            color = PantopusColors.appText,
            modifier =
                Modifier
                    .padding(top = Spacing.s2)
                    .semantics { heading() }
                    .testTag("reportUser_title"),
        )

        Text(
            text =
                "Tell us why you're reporting this account. Our moderators review " +
                    "every report.",
            style = PantopusTextStyle.small,
            color = PantopusColors.appTextSecondary,
        )

        ReasonGroup(
            selected = selectedReason,
            onSelect = onSelectReason,
        )

        DetailsField(
            required = detailsRequired,
            value = details,
            onChange = onDetailsChange,
        )

        if (state is ReportSheetUiState.Failed) {
            Text(
                text = state.message,
                style = PantopusTextStyle.small,
                color = PantopusColors.error,
                modifier = Modifier.testTag("reportUser_error"),
            )
        }

        Spacer(modifier = Modifier.height(Spacing.s1))

        PrimaryButton(
            title = "Submit report",
            onClick = onSubmit,
            isEnabled = canSubmit,
            isLoading = isSubmitting,
            modifier =
                Modifier
                    .fillMaxWidth()
                    .testTag("reportUser_submit"),
        )

        GhostButton(
            title = "Cancel",
            onClick = onCancel,
            isEnabled = !isSubmitting,
            modifier =
                Modifier
                    .fillMaxWidth()
                    .testTag("reportUser_cancelGhost"),
        )
    }
}

@Composable
private fun ReasonGroup(
    selected: ReportReason?,
    onSelect: (ReportReason) -> Unit,
) {
    Column(
        modifier = Modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(Spacing.s2),
    ) {
        Text(
            text = "Reason",
            style = PantopusTextStyle.caption,
            fontWeight = FontWeight.SemiBold,
            color = PantopusColors.appTextSecondary,
            modifier = Modifier.semantics { heading() },
        )
        Column(
            modifier =
                Modifier
                    .fillMaxWidth()
                    .clip(RoundedCornerShape(Radii.lg))
                    .background(PantopusColors.appSurface)
                    .border(
                        width = 1.dp,
                        color = PantopusColors.appBorderSubtle,
                        shape = RoundedCornerShape(Radii.lg),
                    ).testTag("reportUser_reasons"),
        ) {
            val reasons = ReportReason.values().toList()
            reasons.forEachIndexed { index, reason ->
                ReasonRow(
                    reason = reason,
                    isSelected = selected == reason,
                    isLast = index == reasons.lastIndex,
                    onTap = { onSelect(reason) },
                )
            }
        }
    }
}

@Composable
private fun ReasonRow(
    reason: ReportReason,
    isSelected: Boolean,
    isLast: Boolean,
    onTap: () -> Unit,
) {
    Column {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(Spacing.s3),
            modifier =
                Modifier
                    .fillMaxWidth()
                    .heightIn(min = 48.dp)
                    .clickable(onClick = onTap)
                    .padding(horizontal = Spacing.s4)
                    .testTag("reportUser_reason_${reason.key}")
                    .semantics {
                        contentDescription = reason.label
                        selected = isSelected
                    },
        ) {
            Text(
                text = reason.label,
                style = PantopusTextStyle.body,
                color = PantopusColors.appText,
                modifier = Modifier.weight(1f),
            )
            RadioDot(isSelected = isSelected)
        }
        if (!isLast) {
            Box(
                modifier =
                    Modifier
                        .padding(start = Spacing.s4)
                        .fillMaxWidth()
                        .height(1.dp)
                        .background(PantopusColors.appBorderSubtle),
            )
        }
    }
}

@Composable
private fun RadioDot(isSelected: Boolean) {
    Box(
        modifier =
            Modifier
                .size(22.dp)
                .clip(CircleShape)
                .border(
                    width = 1.5.dp,
                    color = if (isSelected) PantopusColors.primary600 else PantopusColors.appBorder,
                    shape = CircleShape,
                ),
        contentAlignment = Alignment.Center,
    ) {
        if (isSelected) {
            Box(
                modifier =
                    Modifier
                        .size(11.dp)
                        .clip(CircleShape)
                        .background(PantopusColors.primary600),
            )
        }
    }
}

@Composable
private fun DetailsField(
    required: Boolean,
    value: String,
    onChange: (String) -> Unit,
) {
    Column(
        modifier = Modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(Spacing.s2),
    ) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(Spacing.s2),
            modifier = Modifier.semantics { heading() },
        ) {
            Text(
                text = "Details",
                style = PantopusTextStyle.caption,
                fontWeight = FontWeight.SemiBold,
                color = PantopusColors.appTextSecondary,
            )
            if (required) {
                Text(
                    text = "Required",
                    style = PantopusTextStyle.caption,
                    color = PantopusColors.error,
                    modifier = Modifier.testTag("reportUser_detailsRequired"),
                )
            }
        }
        Box(
            modifier =
                Modifier
                    .fillMaxWidth()
                    .heightIn(min = 96.dp)
                    .clip(RoundedCornerShape(Radii.md))
                    .background(PantopusColors.appSurface)
                    .border(
                        width = 1.dp,
                        color = PantopusColors.appBorderSubtle,
                        shape = RoundedCornerShape(Radii.md),
                    ).padding(Spacing.s3),
        ) {
            BasicTextField(
                value = value,
                onValueChange = onChange,
                textStyle = PantopusTextStyle.body.copy(color = PantopusColors.appText),
                cursorBrush = SolidColor(PantopusColors.primary600),
                minLines = 4,
                modifier =
                    Modifier
                        .fillMaxWidth()
                        .testTag("reportUser_details")
                        .semantics {
                            contentDescription =
                                if (required) "Details, required" else "Details"
                        },
                decorationBox = { inner ->
                    if (value.isEmpty()) {
                        Text(
                            text =
                                if (required) {
                                    "Tell us what happened"
                                } else {
                                    "Add anything that helps (optional)"
                                },
                            style = PantopusTextStyle.body,
                            color = PantopusColors.appTextMuted,
                        )
                    }
                    inner()
                },
            )
        }
    }
}

File: frontend/apps/ios/Pantopus/Features/Profile/ReportUserSheet.swift
//
//  ReportUserSheet.swift
//  Pantopus
//
//  P6.2 — Report-a-user form, presented as a sheet from the public
//  profile overflow menu. Posts to `POST /api/users/:userId/report`
//  (route `backend/routes/users.js:4153`).
//

import Foundation
import Logging
import Observation
import SwiftUI

/// Reasons surfaced by the design's radio group. Each carries the
/// backend-accepted `reason` key — the Joi schema only accepts a
/// fixed set so design-only categories (impersonation, fraud, hate
/// speech) collapse to a backend bucket and prefix the moderator-visible
/// `details` with their user-facing label.
public enum ReportReason: String, CaseIterable, Identifiable, Sendable {
    case spam
    case harassment
    case impersonation
    case fraud
    case hateSpeech = "hate_speech"
    case other

    public var id: String {
        rawValue
    }

    /// Label rendered in the radio row.
    public var label: String {
        switch self {
        case .spam: "Spam"
        case .harassment: "Harassment"
        case .impersonation: "Impersonation"
        case .fraud: "Fraud"
        case .hateSpeech: "Hate speech"
        case .other: "Other"
        }
    }

    /// Backend-valid key submitted in the `reason` field.
    var backendKey: String {
        switch self {
        case .spam: "spam"
        case .harassment, .hateSpeech: "harassment"
        case .impersonation, .fraud, .other: "other"
        }
    }

    /// Prefix prepended to the moderator-visible `details` body when this
    /// reason is collapsed onto a generic backend key. `nil` means the
    /// backend key already matches the design category.
    var detailsPrefix: String? {
        switch self {
        case .impersonation: "[Impersonation]"
        case .fraud: "[Fraud]"
        case .hateSpeech: "[Hate speech]"
        case .spam, .harassment, .other: nil
        }
    }
}

/// Submission state for the report sheet.
public enum ReportUserSheetState: Sendable, Equatable {
    case idle
    case submitting
    case succeeded
    case failed(message: String)
}

/// View-model behind `ReportUserSheet`. Owns the radio selection, the
/// details field, and the submit lifecycle.
@MainActor
@Observable
public final class ReportUserSheetViewModel {
    public private(set) var state: ReportUserSheetState = .idle
    public var selectedReason: ReportReason?
    public var details: String = ""

    private let userId: String
    private let client: APIClient
    private let logger = Logger(label: "app.pantopus.ios.ReportUser")

    init(userId: String, client: APIClient = .shared) {
        self.userId = userId
        self.client = client
    }

    /// Submit is enabled when a reason is picked and — for `.other` — a
    /// non-empty `details` string is provided. The same rule mirrors on
    /// Android (`ReportUserSheetViewModel#canSubmit`).
    public var canSubmit: Bool {
        guard let selectedReason else { return false }
        if selectedReason == .other {
            return !details.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        }
        return true
    }

    /// Whether the `details` field is gated as required by the current
    /// reason. Drives the "Required" caption + placeholder copy.
    public var detailsRequired: Bool {
        selectedReason == .other
    }

    public func submit() async {
        guard let selectedReason, canSubmit else { return }
        guard state != .submitting, state != .succeeded else { return }
        state = .submitting
        let trimmed = details.trimmingCharacters(in: .whitespacesAndNewlines)
        let body = UserReportBody(
            reason: selectedReason.backendKey,
            details: payloadDetails(for: selectedReason, userDetails: trimmed)
        )
        do {
            _ = try await client.request(
                UserReportsEndpoints.report(userId: userId, body: body),
                as: EmptyResponse.self
            )
            state = .succeeded
        } catch let error as APIError {
            logger.warning("Report submit failed: \(error)")
            state = .failed(message: friendlyMessage(for: error))
        } catch {
            logger.warning("Report submit failed: \(error)")
            state = .failed(message: "Couldn't submit your report.")
        }
    }

    private func payloadDetails(for reason: ReportReason, userDetails: String) -> String? {
        switch (reason.detailsPrefix, userDetails.isEmpty) {
        case (nil, true): nil
        case (nil, false): userDetails
        case let (prefix?, true): prefix
        case let (prefix?, false): "\(prefix) \(userDetails)"
        }
    }

    private func friendlyMessage(for error: APIError) -> String {
        switch error {
        case .notFound: "We couldn't find that user."
        case .forbidden: "You don't have permission to do that."
        case .transport: "Check your connection and try again."
        default: "Couldn't submit your report."
        }
    }
}

/// Sheet content presented from `PublicProfileView` when the overflow's
/// Report row is tapped.
@MainActor
public struct ReportUserSheet: View {
    @State private var viewModel: ReportUserSheetViewModel
    private let handle: String?
    private let displayName: String
    private let onClose: @MainActor () -> Void
    private let onSubmitted: @MainActor () -> Void
    @FocusState private var detailsFocused: Bool

    public init(
        userId: String,
        handle: String?,
        displayName: String,
        onClose: @escaping @MainActor () -> Void,
        onSubmitted: @escaping @MainActor () -> Void
    ) {
        _viewModel = State(initialValue: ReportUserSheetViewModel(userId: userId))
        self.handle = handle
        self.displayName = displayName
        self.onClose = onClose
        self.onSubmitted = onSubmitted
    }

    public var body: some View {
        NavigationStack {
            ScrollView {
                VStack(alignment: .leading, spacing: Spacing.s4) {
                    Text("Tell us why you're reporting this account. Our moderators review every report.")
                        .pantopusTextStyle(.small)
                        .foregroundStyle(Theme.Color.appTextSecondary)

                    reasonGroup
                    detailsField

                    if case let .failed(message) = viewModel.state {
                        Text(message)
                            .pantopusTextStyle(.small)
                            .foregroundStyle(Theme.Color.error)
                            .accessibilityIdentifier("reportUser_error")
                    }

                    actions
                }
                .padding(.horizontal, Spacing.s4)
                .padding(.top, Spacing.s3)
                .padding(.bottom, Spacing.s10)
            }
            .background(Theme.Color.appBg.ignoresSafeArea())
            .navigationTitle(titleText)
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .topBarLeading) {
                    Button("Cancel", action: onClose)
                        .accessibilityIdentifier("reportUser_cancel")
                }
            }
            .accessibilityIdentifier("reportUserSheet")
            .onChange(of: viewModel.state) { _, newState in
                if newState == .succeeded { onSubmitted() }
            }
        }
    }

    private var titleText: String {
        if let handle, !handle.isEmpty { return "Report @\(handle)" }
        return "Report \(displayName)"
    }

    private var reasonGroup: some View {
        VStack(alignment: .leading, spacing: Spacing.s2) {
            Text("Reason")
                .pantopusTextStyle(.caption)
                .fontWeight(.semibold)
                .foregroundStyle(Theme.Color.appTextSecondary)
                .accessibilityAddTraits(.isHeader)
            VStack(spacing: Spacing.s0) {
                ForEach(Array(ReportReason.allCases.enumerated()), id: \.element) { index, reason in
                    ReportReasonRow(
                        reason: reason,
                        isSelected: viewModel.selectedReason == reason,
                        isLast: index == ReportReason.allCases.count - 1
                    ) {
                        viewModel.selectedReason = reason
                    }
                }
            }
            .background(Theme.Color.appSurface)
            .clipShape(RoundedRectangle(cornerRadius: Radii.lg))
            .overlay(
                RoundedRectangle(cornerRadius: Radii.lg)
                    .stroke(Theme.Color.appBorderSubtle, lineWidth: 1)
            )
            .accessibilityIdentifier("reportUser_reasons")
        }
    }

    private var detailsField: some View {
        VStack(alignment: .leading, spacing: Spacing.s2) {
            HStack(spacing: Spacing.s2) {
                Text("Details")
                    .pantopusTextStyle(.caption)
                    .fontWeight(.semibold)
                    .foregroundStyle(Theme.Color.appTextSecondary)
                if viewModel.detailsRequired {
                    Text("Required")
                        .pantopusTextStyle(.caption)
                        .foregroundStyle(Theme.Color.error)
                        .accessibilityIdentifier("reportUser_detailsRequired")
                }
            }
            .accessibilityAddTraits(.isHeader)
            TextField(
                viewModel.detailsRequired
                    ? "Tell us what happened"
                    : "Add anything that helps (optional)",
                text: Binding(
                    get: { viewModel.details },
                    set: { viewModel.details = $0 }
                ),
                axis: .vertical
            )
            .focused($detailsFocused)
            .lineLimit(4...8)
            .font(Theme.Font.body)
            .foregroundStyle(Theme.Color.appText)
            .padding(Spacing.s3)
            .background(Theme.Color.appSurface)
            .clipShape(RoundedRectangle(cornerRadius: Radii.md))
            .overlay(
                RoundedRectangle(cornerRadius: Radii.md)
                    .stroke(Theme.Color.appBorderSubtle, lineWidth: 1)
            )
            .accessibilityIdentifier("reportUser_details")
        }
    }

    @ViewBuilder private var actions: some View {
        let isSubmitting = viewModel.state == .submitting
        VStack(spacing: Spacing.s2) {
            PrimaryButton(
                title: "Submit report",
                isLoading: isSubmitting,
                isEnabled: viewModel.canSubmit && !isSubmitting
            ) {
                await viewModel.submit()
            }
            .accessibilityIdentifier("reportUser_submit")

            GhostButton(title: "Cancel", isEnabled: !isSubmitting) {
                await MainActor.run { onClose() }
            }
            .accessibilityIdentifier("reportUser_cancelGhost")
        }
        .padding(.top, Spacing.s2)
    }
}

/// One tappable row in the reason radio group.
private struct ReportReasonRow: View {
    let reason: ReportReason
    let isSelected: Bool
    let isLast: Bool
    let onTap: @MainActor () -> Void

    var body: some View {
        VStack(spacing: Spacing.s0) {
            Button(action: onTap) {
                HStack(spacing: Spacing.s3) {
                    Text(reason.label)
                        .pantopusTextStyle(.body)
                        .foregroundStyle(Theme.Color.appText)
                    Spacer()
                    RadioDot(isSelected: isSelected)
                }
                .contentShape(Rectangle())
                .padding(.horizontal, Spacing.s4)
                .frame(minHeight: 48)
            }
            .buttonStyle(.plain)
            .accessibilityIdentifier("reportUser_reason_\(reason.rawValue)")
            .accessibilityLabel(reason.label)
            .accessibilityAddTraits(isSelected ? [.isButton, .isSelected] : .isButton)

            if !isLast {
                Divider()
                    .background(Theme.Color.appBorderSubtle)
                    .padding(.leading, Spacing.s4)
            }
        }
    }
}

private struct RadioDot: View {
    let isSelected: Bool

    var body: some View {
        ZStack {
            Circle()
                .stroke(
                    isSelected ? Theme.Color.primary600 : Theme.Color.appBorder,
                    lineWidth: 1.5
                )
                .frame(width: 22, height: 22)
            if isSelected {
                Circle()
                    .fill(Theme.Color.primary600)
                    .frame(width: 11, height: 11)
            }
        }
        .accessibilityHidden(true)
    }
}

#Preview {
    ReportUserSheet(
        userId: "preview",
        handle: "alex",
        displayName: "Alex Rivera",
        onClose: {},
        onSubmitted: {}
    )
}
